rfcs/77-dev-tools-abstraction.md

```diff
@@ -0,0 +1,826 @@
+
+MVP implementation of CLI parser can be found at [CLI-Parser](https://github.com/rewin123/bevy_dev_CLI_prototype/tree/main)
+
+## Implementation strategy
+
+### What metadata do toolboxes need?
+
+We can start simple:
+
+- the type name of the dev tool
+- any doc strings on the dev tool struct: see the [`Documentation` struct](https://github.com/bevyengine/bevy/blob/e33b93e31230c44d3b269d0c781544872fbd3909/crates/bevy_reflect/bevy_reflect_derive/src/documentation.rs#L43) from `bevy_reflect`
+- the value and types of any configuration that must be supplied
+
+By relying on simple structs to configure both our modal dev tools and commands, we can extract this information automatically, via reflection.
+Additional dev tool specific metadata (such as a classification scheme) can be added to the traits on an opt-in basis in the future.
+
+Optional methods on both `ModalDevTool` and `DevCommand` will allow us to override the supplied defaults if needed.
+
+We also need access to one other critical piece of information: a function pointer that allows us to construct a new value of this type from a string.
+
+As a result our `DevToolMetadata` looks like:
+
+```rust
+struct DevToolMetaData {
+    name: &'static str,
+    type_id: TypeId,
+    type_info: &'static TypeInfo,
+    from_str_fn: fn(&str) -> Result<Box<dyn Reflect>, DevToolParseError>,
+    short_description: Option<&'static str>
+}
+```
+
+### What do the registries for our dev tools look like?
+
+Keeping track of the registered dev tools without storing them all in a dedicated collection is quite challenging!
+How can we keep track of it under the hood?
+
+```rust
+#[derive(Resource)]
+struct DevToolsRegistry {
+    /// The stored collection of modal dev tools, tracked in a type-erased way using [`ComponentId`]
+    /// 
+    /// The key is the `name()` provided by the `ModalDevTool` trait.
+    modal_dev_tools: HashMap<String, DevToolMetadata>,
+    /// The metadata for all registered dev commands.
+    /// 
+    /// The key is the `name()` provided by the `DevCommand` trait.
+    dev_commands: HashMap<String, DevCommandMetaData> 
+}
+```
+
+There are a few key operations that we will need to perform:
+
+```rust
+impl DevToolsRegistry {
+    /// Gets a reference to the specified modal dev tool by `ComponentId`.
+    fn get_tool_by_id(world: &World, component_id: ComponentId) -> Option<&dyn ModalDevTool> {
+        let resource = world.get_resource_by_id(component_id)?;
+        resource.downcast_ref()
+    }
+
+    /// Gets a mutable reference to the specified modal dev tool by `ComponentId`.
+    fn get_tool_mut_by_id(mut world: &mut World, component_id: ComponentId) -> Option<&mut dyn ModalDevTool> {
+        let resource = world.get_resource_mut_by_id(component_id)?;
+        resource.downcast_mut()
+    }
```

Comment:
```rust
    pub fn get_tool_by_component_id<'w>(&self, cmp: ComponentId, world: &'w World) -> Option<&'w dyn Reflect> {
        let type_id = world.components().get_info(cmp)?.type_id()?;
        let type_registry = world.resource::<AppTypeRegistry>().read();
        type_registry.get(type_id)?.data::<ReflectResource>()?.reflect(world)
    }

    pub fn get_tool_by_component_id_mut<'w>(&self, cmp: ComponentId, world: &'w mut World) -> Option<Mut<'w, dyn Reflect>> {
        let type_id = world.components().get_info(cmp)?.type_id()?;
        let reflect_resource = {
            let type_registry = world.resource::<AppTypeRegistry>().read();
            a.get(type_id)?.data::<ReflectResource>()?.clone()
        };
        reflect_resource.reflect_mut(world)
    }
```
With this approach, we can return a dyn Reflect instead of dyn ModalDevTool while keeping component_id

Reply:
I don't know for sure the implications of using ComponentId instead of TypeId, this can simplified if we don't need to get the type_id from components first.